File: python/51.N-Queens.py
class Solution:
    def solveNQueens(self, n):
        """
        :type n: int
        :rtype: List[List[str]]
        """
        S = [0] * (n+1)        # 行安全
        U = [0] * (2*n + 1)    # 记录col + row，即向上的对角线是否安全；从左下到右上:1、2...2*n
        D = [0] * (2*n + 1)    # 记录col - row + (n+1), 即向下的对角是否安全；从左下到右上：2*n、2*n-1...3、2、1
        q = [0] * (n+1)
        Q = []
        self.findPosition(q, 1, n, S, U, D, Q)
        return Q

    def findPosition(self, q, col, n, S, U, D, Q):
        if col == n + 1:
            tmpQ = []
            for j in range(1, n+1):
                tmpStr =    '.'*(q[j]-1) + 'Q' + '.'*(n-q[j])
                tmpQ.append(tmpStr)
            Q.append(tmpQ)
            return
            
        for row in range(1, n+1):
            if (S[row] == 0 and U[col+row] == 0 and D[col - row + (n + 1)] == 0):
                q[col] = row
                S[row] = 1
                U[col + row] = 1
                D[col - row + (n + 1)] = 1
                cnt = self.findPosition(q, col+1, n, S, U, D, Q)
                S[row] = 0
                U[col + row] = 0
                D[col - row + (n + 1)] = 0
        return 

if __name__ == '__main__':
    print(Solution().solveNQueens(4))

File: cpp/51.N-Queens.cpp
# include <iostream>
# include <vector>
# include <string>
using namespace std;


class Solution {
public:
    vector<vector<string>> solveNQueens(int n) {
        vector<int> R(n+1);
		vector<int> U(2*n + 1);
		vector<int> D(2*n + 1);
		vector<int> p(n+1);
		vector<vector<string>> result;
		
		for (int i=0; i<n+1; i++)
			R[i] = 0;
		for (int i=0; i<2*n+1; i++){
			U[i] = 0;
			D[i] = 0;
 		}
 		getPosition(1, n, R, U, D, p, result);
 		return result;
    } 
private:
	void getPosition(int col, int n, vector<int>&R, vector<int>&U, vector<int>&D, vector<int>&p, vector<vector<string>> &result){
		if (col == n+1){
			vector<string> tmp(n);
			for (int c=1; c <= n; c++){
				tmp[c-1] = string(n, '.');
				tmp[c-1][p[c]-1] = 'Q';
			}
			result.push_back(tmp);
		}
		for (int row = 1; row <= n; row++){
			if (R[row]==0 && U[col+row]==0 and D[col-row+(n+1)]==0){
				R[row]=1;
				U[col+row]=1;
				D[col-row+(n+1)]=1;
				p[row] = col;               //row和col可以交换，转置矩阵.... 
				getPosition(col+1, n, R, U, D, p, result);
				R[row]=0;
				U[col+row]=0;
				D[col-row+(n+1)]=0;
			}
		}
	}
};
